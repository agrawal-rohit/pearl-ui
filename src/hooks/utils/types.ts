export type AccessibilityRoles =
  | "adjustable"
  | "alert"
  | "button"
  | "checkbox"
  | "combobox"
  | "header"
  | "image"
  | "imagebutton"
  | "keyboardkey"
  | "link"
  | "menu"
  | "menubar"
  | "menuitem"
  | "none"
  | "progressbar"
  | "radio"
  | "radiogroup"
  | "scrollbar"
  | "search"
  | "spinbutton"
  | "summary"
  | "switch"
  | "tab"
  | "tablist"
  | "text"
  | "timer"
  | "togglebutton"
  | "toolbar";

export type MotiProps =
  | "animate"
  | "from"
  | "transition"
  | "delay"
  | "state"
  | "stylePriority"
  | "onDidAnimate"
  | "exit"
  | "exitTransition"
  | "animateInitialState";
