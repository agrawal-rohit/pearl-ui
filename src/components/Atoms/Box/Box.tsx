import React from "react";
import { View, ViewProps } from "react-native";
import { pearlify } from "../../../pearlify";

/**
 * Box is the most abstract component on top of which all other Pearl UI components are built.
 * By default, it renders a <View> element.
 */
const Box = pearlify<ViewProps>(View, {
  componentName: "Box",
  type: "basic",
  animatable: true,
});

// The props that the Box component accepts.
export type BoxProps = React.ComponentProps<typeof Box>;

// Export the Box component as the default export.
export default Box;
