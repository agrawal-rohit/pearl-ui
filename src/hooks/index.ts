export { useColorModeValue } from "./useColorModeValue";
export { useComponentConfig } from "./useComponentConfig";
export { useStyledProps } from "./useStyledProps";
export { useTheme } from "./useTheme";
