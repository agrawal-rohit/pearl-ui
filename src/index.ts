export { default as Box } from "./components/Box/Box";
export { default as Button } from "./components/Button/Button";
export { default as Text } from "./components/Text/Text";
