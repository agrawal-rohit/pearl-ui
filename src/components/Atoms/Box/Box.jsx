import React from "react";
import { View } from "react-native";
import useStyledProps from "../../../hooks/useStyledProps";
import { useTheme } from "../../../hooks/useTheme";
import {
  backgroundColor,
  border,
  layout,
  opacity,
  position,
  shadow,
  spacing,
  visible,
} from "../../../theme/src/styleFunctions";
import {
  IBasePearlTheme,
  RestyleFunctionContainer,
} from "../../../theme/src/types";

// export type BoxProps<
//   Theme extends IBasePearlTheme
// > = BackgroundColorProps<Theme> &
//   OpacityProps &
//   VisibleProps &
//   LayoutProps &
//   SpacingProps<Theme> &
//   BorderProps<Theme> &
//   ShadowProps<Theme> &
//   PositionProps<Theme>;

export const boxRestyleFunctions = [
  backgroundColor,
  opacity,
  visible,
  layout,
  spacing,
  border,
  shadow,
  position,
];

/**
 * Box is the most abstract component on top of which all other Pearl UI components are built. By default, it renders a <View> element.
 */
// type ViewProps = React.ComponentProps<typeof View> & {
//   children?: React.ReactNode;
// };
// type ComponentProps<Theme extends IBasePearlTheme> = BoxProps<Theme> &
//   Omit<ViewProps, keyof BoxProps<Theme>>;

const Box = React.forwardRef((props, ref) => {
  const passedProps = useStyledProps(boxRestyleFunctions, props);

  return <View ref={ref} {...passedProps} />;
});

export default Box;
