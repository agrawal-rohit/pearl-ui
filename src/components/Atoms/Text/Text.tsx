import React from "react";
import { Text as RNText } from "react-native";
import responsiveSize from "../../../utils/responsiveSize";
import { useAtomicComponentConfig } from "../../../hooks/useAtomicComponentConfig";
import {
  backgroundColor,
  BackgroundColorProps,
  color,
  ColorProps,
  layout,
  LayoutProps,
  opacity,
  OpacityProps,
  spacing,
  SpacingProps,
  textShadow,
  TextShadowProps,
  typography,
  TypographyProps,
  visible,
  VisibleProps,
} from "../../../theme/src/styleFunctions";
import {
  ComponentSizes,
  ComponentVariants,
  ResponsiveValue,
  StyleFunctionContainer,
} from "../../../theme/src/types";
import { useTheme } from "../../../hooks/useTheme";

type TextStyleProps = ColorProps &
  BackgroundColorProps &
  OpacityProps &
  VisibleProps &
  LayoutProps &
  TypographyProps &
  SpacingProps &
  TextShadowProps & {
    /** The size of the text */
    size?: ResponsiveValue<ComponentSizes<"Text">>;
    /** The variant of the text */
    variant?: ResponsiveValue<ComponentVariants<"Text">>;
    /** Whether to slightly scale the font size based on the screen dimensions */
    scaleFontSize?: boolean;
  };

export const textStyleFunctions = [
  color,
  backgroundColor,
  opacity,
  visible,
  layout,
  typography,
  spacing,
  textShadow,
] as StyleFunctionContainer[];

type RNTextProps = React.ComponentProps<typeof RNText>;

export type TextProps = TextStyleProps &
  Omit<RNTextProps, keyof TextStyleProps>;

export const buildFontConfig = (
  textStyle: any,
  allowFontScaling: boolean
): object => {
  const fontWeight = textStyle.fontWeight;
  const fontStyle = textStyle.fontStyle || "normal";

  let fontSize = textStyle.fontSize;
  let lineHeight = textStyle.lineHeight;

  if (allowFontScaling) {
    fontSize = responsiveSize(fontSize);
    lineHeight = responsiveSize(lineHeight);
  }

  const initialFontFamily = textStyle.fontFamily;
  const { theme } = useTheme();

  const finalFontFamily =
    theme.fontConfig[initialFontFamily][fontWeight][fontStyle];

  return {
    fontSize,
    fontWeight,
    fontFamily: finalFontFamily,
    fontStyle,
  };
};

/**
 * Text is the component which controls the typography across your app. By default, it renders a <Text /> component
 */
const Text = React.forwardRef((rest: TextProps, ref: any) => {
  const props = useAtomicComponentConfig(
    "Text",
    rest,
    {
      size: rest.size,
      variant: rest.variant,
    },
    "primary",
    textStyleFunctions
  );

  const memoizedBuildFontConfig = React.useCallback(
    () => buildFontConfig(props.style, props.scaleFontSize),
    [props.style, props.scaleFontSize]
  );

  props.style = {
    includeFontPadding: false,
    ...props.style,
    ...memoizedBuildFontConfig(),
  };

  return (
    <RNText
      ref={ref}
      accessible={true}
      accessibilityRole="text"
      allowFontScaling
      {...props}
    >
      {props.children}
    </RNText>
  );
});

export default Text;
