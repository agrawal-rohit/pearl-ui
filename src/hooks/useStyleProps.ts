import { useTheme } from "./useTheme";
import { StyleFunctionContainer } from "../theme/src/types";
import { buildFinalStyleProps, composeStyleProps } from "./utils/utils";
import { useMemo } from "react";
import _ from "lodash";
import { useWindowDimensions } from "react-native";

/**
 * Hook to convert the received style props to appropriate React Native styles
 * @param props Raw props passed to the component where the hook is being used.
 * @param styleFunctions List of style functions to use for computing the received style props.
 * @returns Object containing the computed style properties
 */
export const useStyleProps = (
  props: Record<string, any>,
  styleFunctions: StyleFunctionContainer[]
) => {
  const { theme, colorMode } = useTheme();
  const dimensions = useWindowDimensions();

  // Compose the style functions
  const buildStyleProperties = useMemo(
    () => composeStyleProps(styleFunctions),
    [styleFunctions]
  );

  // If no props are passed, return an empty style object
  if (_.isEmpty(props)) {
    return { style: {} };
  }

  // Build the final style properties
  return buildFinalStyleProps(props, buildStyleProperties, {
    theme,
    colorMode,
    dimensions,
  });
};
