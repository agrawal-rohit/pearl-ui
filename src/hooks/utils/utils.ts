import _ from "lodash";
import { Dimensions, FinalPearlTheme } from "../../theme/src/types";
import { getKeys } from "../../theme/utils/typeHelpers";

export const checkKeyAvailability = (
  key: string,
  object: object,
  objectVarName: string | undefined = undefined
) => {
  if (!object.hasOwnProperty(key)) {
    throw new Error(
      `Key '${key}' does not exist in ${objectVarName ? objectVarName : object}`
    );
  }
};

export const filterStyledProps = (props: any, omitList: any) => {
  const omittedProp = omitList.reduce((acc: any, prop: any) => {
    acc[prop] = true;
    return acc;
  }, {});

  return getKeys(props).reduce((acc: any, key: any) => {
    if (!omittedProp[key]) {
      acc[key] = props[key];
    }
    return acc;
  }, {});
};

export const buildFinalStyleProps = (
  props: Record<string, any>,
  buildStyleProperties: any,
  { theme, dimensions }: { theme: FinalPearlTheme; dimensions: Dimensions }
) => {
  const coreVisualStyle = buildStyleProperties.buildStyle(props, {
    theme,
    dimensions,
  });

  const cleanStyleProps = filterStyledProps(
    props,
    buildStyleProperties.properties
  );

  cleanStyleProps.style = { ...coreVisualStyle, ...props.style };
  return cleanStyleProps;
};

export const composeMotiProps = (
  props: Record<string, any>,
  buildStyleProperties: any,
  { theme, dimensions }: { theme: FinalPearlTheme; dimensions: Dimensions }
) => {
  let motiStyleProps = buildFinalStyleProps(props, buildStyleProperties, {
    theme,
    dimensions,
  });
  motiStyleProps = { ...motiStyleProps, ...motiStyleProps.style };
  motiStyleProps = _.omit(motiStyleProps, ["style", "display", "shadowOffset"]);
  return motiStyleProps;
};
