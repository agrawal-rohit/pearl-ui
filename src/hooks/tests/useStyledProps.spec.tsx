import React, { useState } from "react";
import Text from "../../components/Atoms/Text/Text";
import { fireEvent, render } from "@testing-library/react-native";
import { ThemeProvider } from "../../theme/src/themeContext";
import Box from "../../components/Atoms/Box/Box";
import { useStyleProps } from "../useStyleProps";
import { allStyleFunctions } from "../../theme/src/styleFunctions";
import { getKeys } from "../../theme/utils/typeHelpers";

jest.useFakeTimers();

const TestComponent: React.FC<any> = (props: any) => {
  const rest = useStyleProps(props.object, allStyleFunctions);

  return (
    <Box {...rest}>
      <Text>{getKeys(rest.style).length}</Text>
    </Box>
  );
};

describe("useStyleProps", () => {
  it("return empty style object for missing props", () => {
    const comp = render(
      <ThemeProvider>
        <TestComponent object={null} />
      </ThemeProvider>
    );

    const tree = comp.toJSON();

    expect(tree).toMatchSnapshot();
    expect(comp.getByText("0")).toBeTruthy();
  });
  it("return correct style object for props", () => {
    const comp = render(
      <ThemeProvider>
        <TestComponent object={{ backgroundColor: "pink", p: "s" }} />
      </ThemeProvider>
    );

    const tree = comp.toJSON();

    expect(tree).toMatchSnapshot();
    expect(comp.getByText("3")).toBeTruthy();
  });
});
