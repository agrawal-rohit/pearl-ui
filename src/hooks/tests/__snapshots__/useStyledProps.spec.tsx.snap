// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useStyledProps return correct style object for props 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    animatedStyle={
      Object {
        "value": Object {
          "transform": Array [],
        },
      }
    }
    collapsable={false}
    style={
      Object {
        "backgroundColor": "#ffc0cb",
        "display": "flex",
        "padding": 12,
        "transform": Array [],
      }
    }
  >
    <Text
      accessibilityRole="text"
      accessible={true}
      allowFontScaling={true}
      animatedStyle={
        Object {
          "value": Object {
            "transform": Array [],
          },
        }
      }
      collapsable={false}
      style={
        Object {
          "color": "#1A2138",
          "display": "flex",
          "fontFamily": "Poppins-Regular",
          "fontSize": 18.4,
          "fontStyle": "normal",
          "fontWeight": "400",
          "includeFontPadding": false,
          "letterSpacing": 0,
          "lineHeight": 24,
          "transform": Array [],
        }
      }
    >
      3
    </Text>
  </View>
</View>
`;

exports[`useStyledProps return empty style object for missing props 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    animatedStyle={
      Object {
        "value": Object {
          "transform": Array [],
        },
      }
    }
    collapsable={false}
    style={
      Object {
        "display": "flex",
        "transform": Array [],
      }
    }
  >
    <Text
      accessibilityRole="text"
      accessible={true}
      allowFontScaling={true}
      animatedStyle={
        Object {
          "value": Object {
            "transform": Array [],
          },
        }
      }
      collapsable={false}
      style={
        Object {
          "color": "#1A2138",
          "display": "flex",
          "fontFamily": "Poppins-Regular",
          "fontSize": 18.4,
          "fontStyle": "normal",
          "fontWeight": "400",
          "includeFontPadding": false,
          "letterSpacing": 0,
          "lineHeight": 24,
          "transform": Array [],
        }
      }
    >
      0
    </Text>
  </View>
</View>
`;
